tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -224,14 +241,72 @@ std::vector<TensorView*> mlp(
   return {linear1, gelu, linear2, dropout};
 }
 
+std::vector<TensorView*> mha(
+    TensorView* x,
+    TensorView* w0,
+    TensorView* b0,
+    TensorView* w1,
+    TensorView* b1,
+    DeviceMesh& mesh,
+    DataType dtype) {
+  // Linear 1
+  TensorView* mm = matmul(x, w0);
+  TensorView* proj_bias_bcast = broadcast(b0, {false, true, false});
+  TensorView* qkv1 = add(mm, proj_bias_bcast);
+  // Forming the q,k,v vectors:
+  TensorView* qkv = reshape(qkv1, {D, B * S, 3 * E / D}, {D, B, S, 3 * E / D});
```

Comment:
I believe you can
```
auto D = x->axis(0)->extent()->value().as<int64_t>();
```
to avoid having to read D from the global variable. I always wary of mutable global variables as they make code harder to reason about.